File: src/sprite/currylib/setfunctions.cpp
#include "sprite/builtins.hpp"
#include "sprite/currylib/prelude.hpp"
#include "sprite/currylib/setfunctions.hpp"
#include "sprite/fairscheme.hpp"
#include "sprite/graph/memory.hpp"
#include "sprite/inspect.hpp"

namespace sprite { inline namespace
{
  tag_type allValues_step(RuntimeState * rts, Configuration * C)
  {
    Cursor _0 = C->cursor();
    Variable _1 = variable(_0, 0);
    auto status = rts->hnf(C, &_1);
    if(status != T_CTOR)
      return status;
    ChoiceNode * choice = nullptr;
    SetEvalNode * seteval = NodeU{_1.target}.seteval;
    rts->push_queue(seteval->queue);
    auto value = eval_next(rts);
    assert(value.kind == 'p');
    rts->pop_queue();
    if(value.arg.node->info->tag >= T_CTOR)
    {
      _0->node->forward_to(
          cons(
              value.arg.node
            , Node::create(&allValues_Info, _1.target->node)
            )
        );
      return T_FWD;
    }
    assert(value.arg.node->info->tag == T_CHOICE);
    choice = NodeU{value.arg.node}.choice;
    Configuration * subC = seteval->queue->front();
    assert(subC->root->node == (Node *) choice);
    Queue * Qlhs = seteval->queue;
    Queue * Qrhs = new Queue(seteval->set);
    Node * rhs_seteval = Node::create(seteval->info, seteval->set, Qrhs);
    auto out = Qlhs->begin();
    for(auto p=Qlhs->begin(), e=Qlhs->end(); p!=e; ++p)
    {
      switch((*p)->fingerprint.test(choice->cid))
      {
        case LEFT:         *out++ = *p;
                           break;
        case RIGHT:        Qrhs->push_back(*p);
                           break;
        case UNDETERMINED: *out++ = *p;
                           Qrhs->push_back(*p);
                           break;
      }
    }
    Qlhs->resize(out - Qlhs->begin());
    Node * replacement = make_node<ChoiceNode>(
        choice->cid
      , Node::create(&allValues_Info, (Node *) seteval)
      , Node::create(&allValues_Info, rhs_seteval)
      );
    _0->node->forward_to(replacement);
    return T_FWD;
  }

  tag_type _applyS(RuntimeState * rts, Configuration * C, bool capture)
  {
    Cursor _0 = C->cursor();
    Variable _1 = variable(_0, 0);
    auto status = rts->hnf(C, &_1);
    if(status != T_CTOR)
      return status;
    PartApplicNode * partial = NodeU{_1.target}.partapplic;
    assert(partial->missing >= 1);
    Node * arg = _0->node->successor(1);
    if(!capture)
      arg = Node::create(&SetGuard_Info, nullptr, arg);
    Node * replacement = Node::create(
        &PartialS_Info
      , partial->missing - 1
      , partial->head_info
      , cons(partial->terms, arg)
      );
    _0->node->forward_to(replacement);
    return T_FWD;
  }

  tag_type applyS_step(RuntimeState * rts, Configuration * C)
    { return _applyS(rts, C, false); }

  tag_type captureS_step(RuntimeState * rts, Configuration * C)
    { return _applyS(rts, C, true); }

  // ($##>) f a = (f $>) $## a
  tag_type eagerApplyS_step(RuntimeState * rts, Configuration * C)
  {
    Cursor _0 = C->cursor();
    Node * partial = Node::create(
        &PartApplic_Info
      , 1
      , &eagerApplyS_Info
      , cons(_0->node->successor(0), Nil)
      );
    Node * replacement = Node::create(
        &applygnf_Info, partial, _0->node->successor(1)
      );
    _0->node->forward_to(replacement);
    return T_FWD;
  }

  tag_type evalS_step(RuntimeState * rts, Configuration * C)
  {
    Cursor _0 = C->cursor();
    Variable _1 = variable(_0, 0);
    auto status = rts->hnf(C, &_1);
    if(status != T_CTOR)
      return status;
    PartApplicNode * partial = NodeU{_1.target}.partapplic;
    Set * new_set = new Set();
    Node * goal = partial->materialize();
    auto const arity = goal->info->arity;
    for(index_type i=0; i<arity; ++i)
    {
      Cursor cur = goal->successor(i);
      if(inspect::info_of(cur) == &SetGuard_Info && !inspect::get_set(cur))
        *cur = Node::create(
            &SetGuard_Info, new_set, inspect::get_setguard_value(cur)
          );
    }
    Queue * new_queue = new Queue(new_set, goal);
    Node * seteval = Node::create(&SetEval_Info, new_set, new_queue);
    Node * allvalues = Node::create(&allValues_Info, seteval);
    Node * replacement = Node::create(&Values_Info, allvalues);
    _0->node->forward_to(replacement);
    return T_FWD;
  }

  tag_type exprS_step(RuntimeState * rts, Configuration * C)
  {
    Cursor _0 = C->cursor();
    Node * replacement = Node::create(
        &PartialS_Info
      , Arg(ENCAPSULATED_EXPR)
      , _0->node->successor(0)
      );
    _0->node->forward_to(replacement);
    return T_FWD;
  }

  tag_type set_step(RuntimeState * rts, Configuration * C)
  {
    Cursor _0 = C->cursor();
    Variable _1 = variable(_0, 0);
    auto status = rts->hnf(C, &_1);
    if(status != T_CTOR)
      return status;
    assert(_1.target.info() == &PartApplic_Info);
    _0->node->forward_to(_1.target);
    return T_FWD;
  }

  Node * curry(InfoTable const * fapply, Node * head, Arg * tail, Arg * end)
  {
    while(tail != end)
      head = Node::create(fapply, head, *tail++);
    return head;
  }

  tag_type setN_step(RuntimeState * rts, Configuration * C)
  {
    Cursor _0 = C->cursor();
    index_type const n = _0->node->info->arity - 1;
    Node * setf = Node::create(
        n==0 ? &exprS_Info : &set_Info
      , _0->node->successor(0)
      );
    InfoTable const * fapply = rts->setfunction_strategy == SETF_EAGER
        ? &eagerApplyS_Info : &applyS_Info;
    Node * subexpr = curry(
        fapply, setf, _0->node->begin(), _0->node->end()
      );
    Node * replacement = Node::create(&evalS_Info, subexpr);
    _0->node->forward_to(replacement);
    return T_FWD;
  }
}}

namespace sprite
{
  static InfoTable const * Values_Ctors[] = { &Values_Info };
  Type const Values_Type { Values_Ctors, 1 };

  InfoTable const Values_Info {
      /*tag*/        T_CTOR
    , /*arity*/      1
    , /*alloc_size*/ sizeof(Node1)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "Values"
    , /*format*/     "p"
    , /*step*/       nullptr
    , /*typecheck*/  nullptr
    , /*type*/       &Values_Type
    };

  InfoTable const allValues_Info {
      /*tag*/        T_FUNC
    , /*arity*/      1
    , /*alloc_size*/ sizeof(Node1)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "allValues"
    , /*format*/     "p"
    , /*step*/       allValues_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const applyS_Info {
      /*tag*/        T_FUNC
    , /*arity*/      2
    , /*alloc_size*/ sizeof(Node2)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "applyS"
    , /*format*/     "pp"
    , /*step*/       applyS_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const eagerApplyS_Info {
      /*tag*/        T_FUNC
    , /*arity*/      2
    , /*alloc_size*/ sizeof(Node2)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "$##>"
    , /*format*/     "pp"
    , /*step*/       eagerApplyS_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const captureS_Info {
      /*tag*/        T_FUNC
    , /*arity*/      2
    , /*alloc_size*/ sizeof(Node2)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "captureS"
    , /*format*/     "pp"
    , /*step*/       captureS_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const evalS_Info {
      /*tag*/        T_FUNC
    , /*arity*/      1
    , /*alloc_size*/ sizeof(Node1)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "evalS"
    , /*format*/     "p"
    , /*step*/       evalS_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const exprS_Info {
      /*tag*/        T_FUNC
    , /*arity*/      1
    , /*alloc_size*/ sizeof(Node1)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "exprS"
    , /*format*/     "p"
    , /*step*/       exprS_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set_Info {
      /*tag*/        T_FUNC
    , /*arity*/      1
    , /*alloc_size*/ sizeof(Node1)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set"
    , /*format*/     "p"
    , /*step*/       set_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set0_Info {
      /*tag*/        T_FUNC
    , /*arity*/      1
    , /*alloc_size*/ sizeof(Node1)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set0"
    , /*format*/     "p"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set1_Info {
      /*tag*/        T_FUNC
    , /*arity*/      2
    , /*alloc_size*/ sizeof(Node2)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set1"
    , /*format*/     "pp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set2_Info {
      /*tag*/        T_FUNC
    , /*arity*/      3
    , /*alloc_size*/ sizeof(Node_<3>)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set2"
    , /*format*/     "ppp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set3_Info {
      /*tag*/        T_FUNC
    , /*arity*/      4
    , /*alloc_size*/ sizeof(Node_<4>)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set3"
    , /*format*/     "pppp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set4_Info {
      /*tag*/        T_FUNC
    , /*arity*/      5
    , /*alloc_size*/ sizeof(Node_<5>)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set4"
    , /*format*/     "ppppp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set5_Info {
      /*tag*/        T_FUNC
    , /*arity*/      6
    , /*alloc_size*/ sizeof(Node_<6>)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set5"
    , /*format*/     "pppppp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set6_Info {
      /*tag*/        T_FUNC
    , /*arity*/      7
    , /*alloc_size*/ sizeof(Node_<7>)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set6"
    , /*format*/     "ppppppp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };

  InfoTable const set7_Info {
      /*tag*/        T_FUNC
    , /*arity*/      8
    , /*alloc_size*/ sizeof(Node_<8>)
    , /*typetag*/    NO_FLAGS
    , /*flags*/      NO_FLAGS
    , /*name*/       "set7"
    , /*format*/     "pppppppp"
    , /*step*/       setN_step
    , /*typecheck*/  nullptr
    , /*type*/       nullptr
    };
}

File: src/python/backends/py/currylib/setfunctions.py
from ....common import T_CHOICE, LEFT, RIGHT
from copy import copy
from ..eval import fairscheme
from ...generic.currylib import setfunctions as generic_setfunctions
from ...generic.eval.control import E_UNWIND
from .. import graph
from .... import inspect

SetFunctionsSpecification = generic_setfunctions.SetFunctionsSpecification

NO_SID = -1            # an undetermined set ID.
ENCAPSULATED_EXPR = -1 # indicates a partialS is an encapsulated expression.

def allValues(rts, _0):
  # allValues a :: SetEval sid qid -> [a]
  seteval = rts.variable(_0, 0)
  seteval.hnf()
  sid, qid = seteval.successors
  try:
    with rts.queue_scope(sid=sid, qid=qid):
      try:
        value = next(fairscheme.D(rts))
        yield rts.prelude.Cons
        yield value
        yield graph.Node(rts.setfunctions.allValues, seteval)
      except StopIteration:
        yield rts.prelude.Nil
  except E_UNWIND:
    subconfig = rts.qtable[qid][0]
    assert inspect.tag_of(subconfig.root) == T_CHOICE
    cid = rts.obj_id(config=subconfig)
    with rts.queue_scope(sid=sid, trace=False):
      rts.Q = copy(rts.qtable[qid])
      assert rts.qid != qid
      rhs_qid = rts.qid
      rhs_seteval = graph.Node(seteval.info, sid, rts.qid)
    rts.filter_queue(qid, cid, LEFT)
    rts.filter_queue(rhs_qid, cid, RIGHT)
    yield rts.prelude._Choice
    yield cid
    yield graph.Node(rts.setfunctions.allValues, seteval)
    yield graph.Node(rts.setfunctions.allValues, rhs_seteval)

def applyS(rts, _0, capture=False):
  # applyS :: PartialS (a -> b) -> a -> PartialS b
  partapplic = rts.variable(_0, 0)
  partapplic.hnf()
  missing, term = partapplic.target.successors
  assert inspect.isa_unboxed_int(missing)
  assert inspect.isa_func(term) or inspect.isa_ctor(term)
  arg = _0.target.successors[1]
  assert missing >= 1
  _SetGuard = rts.setfunctions._SetGuard
  if not capture:
    arg = graph.Node(_SetGuard, NO_SID, arg)
  yield rts.setfunctions.PartialS
  yield missing - 1
  yield graph.Node(
      term
    , *(term.successors + [arg])
    , partial=(missing != 1)
    )

def captureS(rts, _0):
  # captureS :: PartialS (a -> b) -> a -> PartialS b
  return applyS(rts, _0, capture=True)

def evalS(rts, _0):
  # evalS :: PartialS a -> Values a
  partapplic = rts.variable(_0, 0)
  partapplic.hnf()
  missing, term = partapplic.target.successors
  sid = rts.create_setfunction()
  if missing == ENCAPSULATED_EXPR:
    goal = term
  else:
    assert missing == 0
    _SetGuard = rts.setfunctions._SetGuard
    goal = graph.Node(
        term.info
      , *[graph.Node(_SetGuard, sid, inspect.get_setguard_value(t))
             if inspect.info_of(t) is _SetGuard.info and inspect.get_set_id(t) == NO_SID
             else t
                 for t in term.successors
           ]
      )
  qid = rts.create_queue(sid, goal)
  yield rts.setfunctions.Values
  yield graph.Node(
      rts.setfunctions.allValues
    , graph.Node(rts.setfunctions.SetEval, sid, qid)
    )

def exprS(rts, _0):
  yield rts.setfunctions.PartialS
  yield ENCAPSULATED_EXPR
  yield _0.successors[0]

def set_(rts, _0):
  # set :: a -> PartialS a
  _1 = rts.variable(_0, 0)
  _1.hnf()
  assert _1.info is rts.prelude._PartApplic.info
  yield rts.setfunctions.PartialS
  yield _1.successors[0]
  yield _1.successors[1]

def setN(rts, _0):
  # setN :: (a1 -> ... -> aN -> b) -> a1 -> ... -> aN -> Values b
  n = len(_0.successors) - 1
  pure = getattr(rts.setfunctions, 'exprS' if n == 0 else 'set')
  setf = graph.Node(pure, _0.successors[0])
  fapply = getattr(
      rts.setfunctions
    , '$##>' if rts.setfunction_strategy == 'eager' else '$>'
    )
  yield rts.setfunctions.evalS
  yield graph.utility.curry(rts, setf, *_0.successors[1:], fapply=fapply)

METADATA = {
    'allValues': {'py.rawfunc': allValues}
  , 'applyS'   : {'py.rawfunc': applyS}
  , 'captureS' : {'py.rawfunc': captureS}
  , 'evalS'    : {'py.rawfunc': evalS}
  , 'exprS'    : {'py.rawfunc': exprS}
  , 'set'      : {'py.rawfunc': set_}
  , 'set0'     : {'py.rawfunc': setN}
  , 'set1'     : {'py.rawfunc': setN}
  , 'set2'     : {'py.rawfunc': setN}
  , 'set3'     : {'py.rawfunc': setN}
  , 'set4'     : {'py.rawfunc': setN}
  , 'set5'     : {'py.rawfunc': setN}
  , 'set6'     : {'py.rawfunc': setN}
  , 'set7'     : {'py.rawfunc': setN}
  }

for f in generic_setfunctions.FUNCTIONS:
  if f.name in METADATA:
    f.update_metadata(METADATA[f.name])
